Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx

```diff
@@ -523,58 +446,80 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
+      mTrapSimulator[trapindex].init(mTrapConfig, digit->getDetector(), digit->getROB(), digit->getMCM());
     }
-    int adc = 20 - (pad % 18) - 1;
     std::vector<o2::MCCompLabel> tmplabels;
     auto digitslabels = digitMCLabels.getLabels(digitcounter);
     for (auto& tmplabel : digitslabels) {
       tmplabels.push_back(tmplabel);
     }
     LOG(debug) << "tmplabels for set data : " << tmplabels.size() << " and gslspan digitlabels size of : " << digitslabels.size();
-    LOG(debug) << " setting data with pad=" << pad << " ti=" << trapindex + 1;
-    mTrapSimulator[trapindex].setData(adc, digit->getADC(), tmplabels);
+    LOG(debug) << " setting data with pad=" << digit->getPad() << " ti=" << trapindex + 1;
+    mTrapSimulator[trapindex].setData(digit->getChannel(), digit->getADC(), tmplabels);
 
-    // now take care of the case of shared pads (the whole reason for doing this pad row wise).
-
-    if (pad % 18 == 0 || (pad + 1) % 18 == 0) { //case of pad 18 and 19 must be shared to preceding trap chip adc 1 and 0 respectively.
-      adc = 20 - (pad % 18) - 1;
-      if (trapindex != 0) { // avoid the case of the first trap chip
-        LOG(debug) << " setting data preceding with pad=" << pad << " ti=" << trapindex - 1;
-        mTrapSimulator[trapindex - 1].setData(adc, digit->getADC(), tmplabels);
-      }
+    digitcounter++;
+  } // end of loop over digits.
+  // don't forget about the last pad row..
+  for (int trapcounter = 0; trapcounter < 8; trapcounter++) {
+    if (!mTrapSimulator[trapcounter].checkInitialized()) {
+      LOG(debug) << "TRAP chip not initialized, as there is no input data set. Skipping it.";
+      continue;
+    }
+    if (!mTrapSimulator[trapcounter].isDataSet()) {
+      LOG(debug) << "TRAP chip initialized, but no data for this pad row. Skipping it.";
+      continue;
     }
-    if ((pad - 1) % 18 == 0) { // case of pad 17 must shared to next trap chip as adc 20
-                               //check trap is initialised.
-      adc = 20 - (pad % 18) - 1;
-      if (trapindex + 1 != 8) { // avoid the case of the last trap chip.
-        LOG(debug) << " setting data proceeding with pad=" << pad << " ti=" << trapindex + 1;
-        mTrapSimulator[trapindex + 1].setData(adc, digit->getADC(), tmplabels);
+    countProcessedTraps++;
+    //this one has been filled with data for the last pad row.
+    auto trapsimtimerstart = std::chrono::high_resolution_clock::now();
+    mTrapUsedCounter[trapcounter]++;
+    mTrapSimulator[trapcounter].filter();
+    mTrapSimulator[trapcounter].tracklet();
+
+    trapTracklets = mTrapSimulator[trapcounter].getTrackletArray64(); //TODO remove the copy and send the Accumulated array into the Trapsimulator
+    auto trapLabels = mTrapSimulator[trapcounter].getTrackletLabels();
+    if (!mNewTrackletHCHeaderHasBeenWritten && trapTracklets.size() != 0) { // take account of the case where we have data in the trapchip adc but no tracklets OS: Is this a TODO?
```

Comment:
not a todo, an explanation.

Reply:
but entire if and else can go. its now buried in raw generation.i.e. up to line 511

Reply:
Well, the comment is confusing with `if trapTracklets.size() !=0` and it says to take account of the case with no tracklets. But ok, will remove the block

Reply:
the header still has to be added to the index even if nothing there.

Reply:
So should it be `trapTracklets.size() == 0` then? But anyhow, the block is removed now